pkg/zpdb/pod_observer.go

```diff
@@ -0,0 +1,109 @@
+// The PodObserver is used by the pod eviction web hook handler.
+type PodObserver struct {
+	podsFactory   informers.SharedInformerFactory
+	podLister     corelisters.PodLister
+	podsInformer  cache.SharedIndexInformer
+	PodEvictCache *PodEvictionCache
+	logger        log.Logger
+	stopCh        chan struct{}
+}
+
+func NewPodObserver(kubeClient kubernetes.Interface, namespace string, logger log.Logger) *PodObserver {
+	namespaceOpt := informers.WithNamespace(namespace)
+
+	// initialize the ZoneAwarePodDisruptionBudget custom resource watching
+	podsFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, informerSyncInterval, namespaceOpt)
+	podsInformer := podsFactory.Core().V1().Pods()
+
+	c := &PodObserver{
+		podsFactory:   podsFactory,
+		podLister:     podsInformer.Lister(),
+		podsInformer:  podsInformer.Informer(),
+		PodEvictCache: NewPodEvictionCache(),
+		logger:        logger,
+		stopCh:        make(chan struct{}),
+	}
+
+	return c
+}
+
+func (c *PodObserver) Start() error {
+	_, err := c.podsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onPodAdded,
+		UpdateFunc: c.onPodUpdated,
+		DeleteFunc: c.onPodDeleted,
+	})
+	if err != nil {
+		return err
+	}
+
+	go c.podsFactory.Start(c.stopCh)
+
+	// Wait until all informer caches have been synced.
+	level.Info(c.logger).Log("msg", "zpdb pod informer caches are syncing")
+	if ok := cache.WaitForCacheSync(c.stopCh, c.podsInformer.HasSynced); !ok {
+		return errors.New("zpdb pod informer failed to sync")
+	}
+	level.Info(c.logger).Log("msg", "zpdb pod informer caches have synced")
+
+	return nil
+}
+
+func (c *PodObserver) invalidatePodEvictionCache(obj interface{}) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		level.Warn(c.logger).Log("msg", "unexpected object passed through informer", "type", reflect.TypeOf(obj))
+		return
+	}
+
+	// reduce logging noise as this code path will be run on any pod update
+	// this is cheaper than finding the zpdb config for a pod
+	// and worst case if we miss an eviction cache removal it self-expires
+	if !c.PodEvictCache.HasPendingEviction(pod) {
+		return
+	}
+
+	// after an eviction request is allowed, the informer observes a pod update which can show it still ready/running
+	// if another pod eviction request comes in before the first eviction takes effect this can incorrectly allow this later eviction request to proceed
+	if util.IsPodRunningAndReady(pod) {
+		level.Info(c.logger).Log("msg", "ignorning pod informer update - pod is still reporting as ready and running", "name", pod.GetName(), "reason", pod.Status.Reason, "phase", pod.Status.Phase, "ready", util.IsPodRunningAndReady(pod))
+		return
+	}
+
+	level.Info(c.logger).Log("msg", "accepting pod informer update - invaliding pod eviction cache", "name", pod.GetName(), "reason", pod.Status.Reason, "phase", pod.Status.Phase, "ready", util.IsPodRunningAndReady(pod))
+	c.PodEvictCache.Delete(pod)
```

Comment:
Might be useful to log the pod's generation here and what it was when we get the eviction request - this might help us piece together the order of events when debugging this.